### PR TITLE
Add example using Sass

### PR DESCRIPTION
**Pull request by eliot-akira**
This is an example of configuring `razzle.config.js` to support Sass, based on a gist mentioned in #297.

### CODE REVIEW

**Comment by vxna** [comment]
Is there any additional configuration required to get it to work with .sass files?

Adding this:

     test: /\.(sass|scss)$/,

Doesn't help and produce an error on compile: 

    ERROR  Failed to compile SERVER with 1 errors
    ./src/styles/App.sass
    Module build failed: Missed semicolon (2:11)

      1 | body
    > 2 |   margin: 0
        |           ^
      3 |   padding: 0
      4 |   font-family: sans-serif
      5 |

**Comment by vxna** [comment]
Sorted, https://github.com/jaredpalmer/razzle/issues/315

**Comment by sebmor** [comment]
If you do 
```
// Comment here
@import 'scsspartial';
```

The build fails with `Unknown word`

It fails on the server

**Comment by paramaggarwal** [comment]
@eliot-akira try using `sass-loader` instead of the `fast-sass-loader` used here. (the original gist in #297 also uses `sass-loader`.)

File: examples/with-scss/razzle.config.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        fallback: 'style-loader',
+        use: [
+          {
+            loader: 'css-loader',
+            options: {
+              importLoaders: 1,
+            },
+          },
+          postCssLoader,
+          'fast-sass-loader',
```
Comment (paramaggarwal) [inline examples/with-scss/razzle.config.js]: please use `sass-loader` for max compatibility

Comment (jaredpalmer) [reply examples/with-scss/razzle.config.js]: Use `sass-loader` por favor.

Comment (eliot-akira) [reply examples/with-scss/razzle.config.js]: Got it. :)

**Comment by mhuggins** [comment]
~~I've tried using this on my new razzle app, but I'm getting an error:~~

>     Module parse failed: /Users/mhuggins/Development/neo/explorer/src/common/styles/app.scss Unexpected token (1:5)
>     You may need an appropriate loader to handle this file type.
>     | body {
>     |   background-color: blue;
>     | }

My `app.scss` file simply contains this:

```scss
body {
  background-color: blue;
}

.app-component {
  background-color: red;
}
```

~~It seems like my razzle.config.js file isn't even being picked up for some reason.~~

**EDIT**: Turns out I accidentally had the razzle.config.js file in my src path. 😞

File: examples/with-scss/razzle.config.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
+// Based on: https://gist.github.com/jaredpalmer/0a91a7bd354b875b913c74f4b16125f7
+
+const path = require('path')
```
Comment (mhuggins) [inline examples/with-scss/razzle.config.js]: This isn't being used.

**Comment by paramaggarwal** [comment]
This is needed soon - even the `create-react-app` team is realising that default SCSS support is needed by popular demand.

(ref: https://twitter.com/dan_abramov/status/917147209170341889)

File: examples/with-scss/razzle.config.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      // On the server, we simply use css-loader to deal with scss imports
+      isServer ? 'css-loader' :
+      // For development, include source map
+      dev
+      ? [
+        'style-loader',
+        {
+          loader: 'css-loader',
+          options: {
+            modules: false,
```
Comment (paramaggarwal) [inline examples/with-scss/razzle.config.js]: I just checked and `modules` is `false` by default. So no need. I understand that the original gist from @jaredpalmer had this - but actually not needed.

File: examples/with-scss/razzle.config.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        'sass-loader',
+      ]
+      // For production, extract CSS
+      : ExtractTextPlugin.extract({
+        fallback: 'style-loader',
+        use: [
+          {
+            loader: 'css-loader',
+            options: {
+              importLoaders: 1,
```
Comment (paramaggarwal) [inline examples/with-scss/razzle.config.js:60]: This means that when we `@import` inside an SCSS file, only `post-css-loader` will run on it and not `sass-loader` - is this the correct expectation? I have been trying to understand the implications. I understand that the original gist by @jaredpalmer also has this - but we could add a comment here about what does this actually do.

Comment (jaredpalmer) [reply examples/with-scss/razzle.config.js:60]: I copied this sass setup from nytimes/kyt should look at their latest release and copy.

**Comment by mhuggins** [comment]
@paramaggarwal Here's an example of a setup that is currently working for me in a production app: https://github.com/neoverse/explorer/blob/master/config/razzle-sass.js

You can see the styles being applied [here](http://explorer.neoverse.io).

**Comment by jaredpalmer** [comment]
@paramaggarwal 

@mhuggins solution is what i've been using in production for 7months, works fine. Confused about what's the issue is?

**Comment by mhuggins** [comment]
Not sure what the issue is, I was just sharing a slightly modified version of this that is working for me.  My solution is derived from this PR with just a little cleanup.

**Comment by jaredpalmer** [comment]
btw you don't even need sass loader on the server. just change css loader's test to include `/.scss$/` and it should be faster.

File: examples/with-scss/razzle.config.js (hunk 4 above)
Comment (paramaggarwal) [reply examples/with-scss/razzle.config.js:60]: I read up more on this - what we have here is correct. It basically tells `css-loader` to use `postcss-loader` on all `@import` lines that it comes across in CSS.

**Comment by mhuggins** [comment]
Hmm, I changed it in [this PR](https://github.com/neoverse/explorer/commit/4aac1fe9eebb539d067cbf490ee7ee8a41cb6fd0#diff-080668f2db4ff077b131fc66d6499954), thought I can't remember why.  I do recall I had some kind of issue before adding that.  I'll play around when I have time and add a comment here for why I did so.

**Comment by paramaggarwal** [comment]
All, I think this PR is ready to go and we should merge it in. My comments are just out of curiosity and not as review comments.

@jaredpalmer It gave an error if I use only `css-loader` on the server, so I initially went with using `['css-loader', 'sass-loader']` - but just today, I have changed it to simply use `ignore-loader`.

If you would like to make source maps work for the SASS files, you need to do what I do in the config that I am using: https://gist.github.com/paramaggarwal/e459f177eba98b67c2fa7f26a202d146 which is to give `sourceMap: dev` as an option to all three loaders.

**Comment by d4rky-pl** [comment]
FYI: ignoring CSS/SASS loaders on server side when using CSS modules (manually with `:local`, not on every file with `modules: true`) will result in FOUC (Flash Of Unstyled Content) in React 15 and may result in broken render in React 16 due to difference in generated class names. 

It bit us already in a project once.

**Comment by jaredpalmer** [comment]
Only during dev

**Comment by Grsmto** [comment]
Having the same issue as @sebmor when using @import . Anyone figured out where it comes from?
Seems to kinda work when I delete the declaration and HMR kicks in...

File: examples/with-scss/razzle.config.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+          })
+        ]
+      }
+    }
+
+    appConfig.module.rules.push({
+      test: /.scss$/,
+      use:
+      // On the server, we simply use css-loader to deal with scss imports
+      isServer ? 'css-loader' :
```
Comment (Grsmto) [inline examples/with-scss/razzle.config.js]: Replacing this line by `isServer ? ['css-loader', 'sass-loader'] :` seems to fix the `@import` not working in dev as commented there https://github.com/jaredpalmer/razzle/pull/314#issuecomment-323797303

Comment (mhuggins) [reply examples/with-scss/razzle.config.js]: Yep, this is what I'm doing in my app as well.

https://github.com/neoverse/explorer/blob/4aac1fe9eebb539d067cbf490ee7ee8a41cb6fd0/config/razzle-sass.js#L26

File: examples/with-scss/razzle.config.js (hunk 3 above)
Comment (eliot-akira) [reply examples/with-scss/razzle.config.js]: OK, removed.

File: examples/with-scss/razzle.config.js (hunk 2 above)
Comment (eliot-akira) [reply examples/with-scss/razzle.config.js]: OK, removed.

File: examples/with-scss/razzle.config.js (hunk 5 above)
Comment (eliot-akira) [reply examples/with-scss/razzle.config.js]: Sure, I updated this to use both `css-loader` and `sass-loader` on the server.

**Comment by Grsmto** [comment]
@eliot-akira hey, dev works fine now. But `npm run build` is breaking with `You may need an appropriate loader to handle this file type.`.

**Comment by eliot-akira** [comment]
Just added sourcemaps for all loaders, as per [comment](#issuecomment-339698419) by @paramaggarwal.

@Grsmto Hmm, I'm not seeing that on my end - the build completes as usual.

**Comment by Grsmto** [comment]
@eliot-akira my bad. It's working fine now!

**Comment by dmitru** [comment]
Any updates for this one? What is stopping from merging this PR? 🤔

File: examples/with-scss/razzle.config.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
+// Based on: https://gist.github.com/jaredpalmer/0a91a7bd354b875b913c74f4b16125f7
+
+const autoprefixer = require('autoprefixer')
+const ExtractTextPlugin = require('extract-text-webpack-plugin')
```
Comment (hrasoa) [inline examples/with-scss/razzle.config.js:4]: Shouldn't we use [mini-css-extract-plugin](https://github.com/webpack-contrib/mini-css-extract-plugin) ?

File: examples/with-scss/razzle.config.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+              importLoaders: 1,
+            },
+          },
+          postCssLoader,
+          'sass-loader',
+        ],
+      })
+    })
+
+    if (!isServer && !dev) {
```
Comment (hrasoa) [inline examples/with-scss/razzle.config.js:69]: If you use [mini-css-extract-plugin](https://github.com/webpack-contrib/mini-css-extract-plugin), you won't need this anymore as ```appConfig.plugins``` already contains the plug in set up:
https://github.com/jaredpalmer/razzle/blob/560a9f490bdcac89b29a03b7d9c471ff21ff6619/packages/razzle/config/createConfig.js#L496

**Pull request closed without merging**